File: VoightKampff/VoightKampff.cpp
// Voight Kampff C++ implementation
// Octavi Font
// http://olimpiada-informatica.org/?cmd=problema&pbm=voightkampff

#include <iostream>
#include <map>
#include <vector>
#include <set>
#include <utility>
#include <string>

using namespace std;

typedef std::map<string, bool> BMAP;
typedef std::map<string, pair<bool, bool> > BBMAP;
typedef std::map<string, short int> IMAP;
typedef std::map<string, BMAP > DMAP;
typedef std::map<string, BBMAP > DDMAP;


class Graph
{
public:
    DDMAP nodes;
    DMAP edgeValues;
    IMAP human;
    bool isConsistent;

    Graph() :
        isConsistent(true)
    {
    }

    void addNode(const string &id) {
        nodes[id] = BBMAP();
        edgeValues[id] = BMAP();
        human[id] = -1;
    };

    void addEdge(const string &source, const string &dest, bool isHuman) {
        if (edgeValues[source].find(dest) != edgeValues[source].end())
        {
            if (edgeValues[source][dest] != isHuman)
                isConsistent = false;
        }
        if (edgeValues[dest].find(source) != edgeValues[dest].end())
        {
            if (edgeValues[dest][source] != isHuman)
                isConsistent = false;
        }
        edgeValues[source][dest] = isHuman;
        nodes[source][dest] = make_pair(isHuman, true);
        nodes[dest][source] = make_pair(isHuman, false);
    }

    void delEdge(const string &source, const string &dest) {
        nodes[source].erase(dest);
        nodes[dest].erase(source);
    }

};

short int nodeValue(const short int node, const pair<bool, bool> pair) {
    if (pair.second)
    {
        if (node == 1) // source is human, therefore, dest is what source says
            return pair.first * 1;
        else // source is replicant, therefore dest is what source does NOT say
            return !pair.first * 1;
    }
    else { // If it is said
        if (node == pair.first) // What was said is true, source is human
            return 1;
        else // What was said is false, source is replicant
            return 0;
    }
}


vector<string> dfs(Graph &g, string &nodeStart) {
    vector<string> traversedNodes;
    vector<string> stack;
    stack.push_back(nodeStart);
    g.human[nodeStart] = 1;
    while(!stack.empty()) {
        string node = stack.back();
        short int nodeIsHuman = g.human[node];
        stack.pop_back();
        traversedNodes.push_back(node);
        vector<string> neighbours;
        vector<pair<string, string> > deletions;
        for(BBMAP::iterator it = g.nodes[node].begin(); it != g.nodes[node].end(); it++) {
            string neighbour = it->first;
            if (g.human[neighbour] == -1)
                g.human[neighbour] = nodeValue(nodeIsHuman, it->second);
            else {
                bool inconsistency;
                inconsistency = g.human[neighbour] != nodeValue(nodeIsHuman, it->second);
                if (inconsistency)
                    throw false;
            }
            deletions.push_back(make_pair(node, neighbour));
            stack.push_back(neighbour);
        }
        for (vector<pair<string, string> >::iterator it = deletions.begin(); it != deletions.end(); ++it)
            g.delEdge(it->first, it->second);
    }
    return traversedNodes;
}

int voightKampff(Graph &g) {
    if (!g.isConsistent)
        return 0;
    set<string> nodeIds;
    int combinations = 1;
    for (DDMAP::iterator it = g.nodes.begin(); it != g.nodes.end(); it++) {
        nodeIds.insert(it->first);
    }
    while (!nodeIds.empty()) {
        vector<string> results;
        string node = *nodeIds.begin();
        try {
            results = dfs(g, node);
        }
        catch (bool b) {
                return 0;
        }
        combinations *= 2;
        for (vector<string>::iterator it = results.begin(); it != results.end(); ++it)
            nodeIds.erase(*it);
    }
    return combinations;
}

int main() {
    typedef map<string, short int> Subjects;
    typedef std::map<string, std::vector<pair<string, bool> > > Statements;
    typedef std::map<string, bool> Checked;

    short int numSubjects;
    short int caseNum = 1;
    while(cin >> numSubjects) {
        Graph g;
        Subjects isHuman;
        Statements statements;
        Checked isChecked;
        // -1: not initialized; 0: replicant, 1: human
        for (int i = 0; i < numSubjects; ++i) {
            string subject;
            cin >> subject;
            g.addNode(subject);
        }
        int numStatetements;
        cin >> numStatetements;
        for (int i = 0; i < numStatetements; ++i)
        {
            string subject, s, referee, thing;
            cin >> subject >> s >> referee >> s >> thing;
            if (thing == "a") {// This is a replicant
                cin >> s;
                g.addEdge(subject, referee, false);
            }
            else // This is a human
                g.addEdge(subject, referee, true);
        }
        int combinations = voightKampff(g);
        cout << "Case " << caseNum << ": " << combinations << endl;
        caseNum++;
    }
}

File: VoightKampff/VoightKampff.py
#!/usr/bin/env python
# encoding: utf-8
"""
Voight Kampff python implementation
http://olimpiada-informatica.org/?cmd=problema&pbm=voightkampff

Olimpiada Matemática Española, Graphs section
"""
import sys

subjects = None
graph = None


class Graph(object):
    nodes = {}  # Dict with node id (name) and dict with neighbours
    human = {}  # Init node values to none by default
    edge_values = {}
    is_consistent = True

    def add_node(self, id):
        self.nodes[id] = {}
        self.human[id] = None
        self.edge_values[id] = {}

    def add_edge(self, source, dest, isHuman):
        # Special case:
        # subject statement inconsistent with previous statement of the same subject.
        if dest in self.edge_values[source]:
            if self.edge_values[source][dest] is not isHuman:
                self.is_consistent = False
        # Special case:
        # subject statement inconsistent with another subject statement (if they accuse each other,
        # they must report the same).
        if source in self.edge_values[dest]:
            if not isHuman is self.edge_values[dest][source]:
                self.is_consistent = False

        self.edge_values[source][dest] = isHuman

        def says():
            if self.human[source]:  # If source is human, destination is whatever source says
                return isHuman
            else:  # Otherwise, destination is the contrary to what source says it is
                return not isHuman

        def is_said():
            return self.human[dest] is isHuman  # If source has not lied, source is human

        self.nodes[source][dest] = says
        self.nodes[dest][source] = is_said

    def del_edge(self, source, dest):
        del self.nodes[source][dest]
        del self.nodes[dest][source]


def parse():
    global subjects
    global graph
    line = sys.stdin.readline()
    data = line.split()
    numSubjects = int(data[0])
    numStatements = int(data[numSubjects + 1])
    subjects = data[1:numSubjects + 1]
    graph = Graph()
    for subject in subjects:
        graph.add_node(subject)
    for x in xrange(numStatements):
        line = sys.stdin.readline()
        data = line.split()
        is_human = None
        if data[-1] == "human":
            is_human = True
        else:
            is_human = False
        graph.add_edge(data[0], data[2], is_human)
    sys.stdin.readline()  # Blank line, ignore


def dfs(node):
    """Return a list of traversed nodes or False if Graph not satisfiable"""
    global graph
    graph.human[node] = True
    traversed_nodes = set()
    stack = []
    stack.append(node)
    while stack:
        node = stack.pop()
        traversed_nodes.add(node)
        neighbours = graph.nodes[node].keys()
        for neighbour in neighbours:
            if graph.human[neighbour] is None:
                graph.human[neighbour] = graph.nodes[node][neighbour]()
            else:
                inconsistency = graph.human[neighbour] is not graph.nodes[node][neighbour]()
                if inconsistency:
                    return False
            graph.del_edge(node, neighbour)
            stack.append(neighbour)
    return traversed_nodes


def voight_kampff():
    global subjects
    global graph
    if not graph.is_consistent:
        return 0
    nodes = set(subjects)
    connected_components = 0
    while nodes:
        node = iter(nodes).next()
        traversed_nodes = dfs(node)
        if not traversed_nodes:
            return 0
        else:
            connected_components += 1
            nodes -= traversed_nodes
    return 2**connected_components


if __name__ == '__main__':
    counter = 1
    while True:
        try:
            parse()
            print "Case {0}: {1}".format(counter, voight_kampff())
            counter += 1
        except:
            break
